envoy/network/connection_handler.h

```diff
@@ -133,8 +155,11 @@ class ConnectionHandler {
 
     /**
      * Stop listening according to implementation's own definition.
+     * @param options can be dynamically cast to provide any extra options that some
+     *                subset of listeners might use, e.g. Quic listeners may need
+     *                to configure packet forwarding during hot restart.
      */
-    virtual void shutdownListener() PURE;
+    virtual void shutdownListener(ExtraShutdownListenerOptionsSharedPtr options = nullptr) PURE;
```

Comment:
Instead of having to do a dynamic_cast (which is not supported on all Envoy build environments) could we define a method that all Options are required to implement (onPacket(...) or some such) and just call that?

Reply:
Hm, it seems like dynamic_cast is pretty much already a requirement (crypto.h and conn_pool.h for example use dynamic_casts), but okay, I changed it to a way that doesn't use dynamic cast *and* still keeps it flexible for future behaviors.

Reply:
If I understand things correctly, in some environments, dynamic_cast is transformed to (essentially) static_cast. So it will still cast from type A to type B, but it does not have the "return null if the classes are incompatible" logic. That means code like:
`ASSERT(dynamic_cast<T*>(&context) != nullptr);`
"works", but is basically a no-op.

I *think*.

Reply:
Ah, I see, and with no RTTI dynamic_cast becomes a compile-time error if you're casting between types that don't have a known direct relation (but then why do we use dynamic_cast rather than static_cast in those contexts!)

Reply:
I think it turns into undefined behavior at runtime! Let's say we have 3 classes, Base, DerivedA and DerivedB. Code will compile just fine which does:
`dyanmic_cast<DerivedA*>(base_ptr);` or `dyanmic_cast<DerivedA*>(base_ptr);`
However, if base_ptr is actually a pointer to a DerivedB object then we get a behavior change. dynamic_cast will return nullptr (because the RTTI lets dynamic_cast know that the object in question is NOT DerivedA.). However, static cast will just blindly return a totally-bogus DerivedA* pointer to a DerivedB object. This is ... bad :)